src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -812,13 +812,50 @@ AdjustDistributedExecutionAfterLocalExecution(DistributedExecution *execution)
 
 /*
  * ExecuteUtilityTaskListWithoutResults is a wrapper around executing task
- * list for utility commands. It simply calls in adaptive executor's task
- * execution function.
+ * list for utility commands. For remote tasks, it simply calls in adaptive
+ * executor's task execution function. For local tasks (if any), kicks Process
+ * Utility via CitusProcessUtility for utility commands. As some local utility
+ * commands can trigger udf calls, this function also processes those udf calls
+ * locally.
  */
 void
-ExecuteUtilityTaskListWithoutResults(List *taskList)
+ExecuteUtilityTaskListWithoutResults(List *taskList, bool localExecutionSupported)
 {
-	ExecuteTaskList(ROW_MODIFY_NONE, taskList, MaxAdaptiveExecutorPoolSize);
+	RowModifyLevel rowModifyLevel = ROW_MODIFY_NONE;
+
+	List *localTaskList = NIL;
+	List *remoteTaskList = NIL;
+
+	/*
+	 * Divide tasks into two if localExecutionSupported is set to true and execute
+	 * the local tasks
+	 */
+	if (localExecutionSupported && ShouldExecuteTasksLocally(taskList))
```

Comment:
I wonder whether we should expand ShouldExecuteTasksLocally a bit. For instance, it could make sense to always execute locally when in a transaction block, to make sure we support cases where a subsequent command requires local execution (e.g. a join between a local table and a reference table).

Reply:
That's a hard call to make. (Similarly, we could consider switching to sequential mode if a reference table with fkey is modified in a transaction block.)

I'm not against it, but it'd definitely change the MX experience. We could maybe consider doing it only for reference tables.

Reply:
Maybo not a common use case (for now) but it would also affect coordinator `shouldhaveshards=true` experience

Reply:
Actually, the further improvement in my is to enable local execution for any local placement (not only for single shard ones -read reference tables-). But, as that would break the parallelism citus provides in a transaction block, doing some bookkeeping for the placements accessed locally so we just enforce local execution for the placements accessed locally previously (not for all the subsequent commands operating on other distributed tables).

Reply:
I'm also wondering why we can't perform local execution in parallel. I really don't know if we can benefit parallelism without opening new connections. If we could perform local execution parallely, we won't be thinking of those cases, we can just locally execute local tasks and the others via remote connections etc.

Reply:
Postgres is gradually implementing parallelism for other commands, but iirc so far only SELECT and CREATE INDEX are supported. At some point, it may be worth exploring whether we could leverage that.